docs/architecture/ADR-002-universal-share-encoding.md

```diff
@@ -0,0 +1,144 @@
+# ADR 009: Universal Share Encoding
+<!-- disable markdownlint MD010 because Go code snippet uses tabs -->
+<!-- markdownlint-disable MD010 -->
+
+## Terminology
+
+- **nid** (8 bytes): namespace id
+- **reserved** (1 byte): is the location of the first transaction, ISR, or evidence in this share if there is one and `0` if there isn't one
+- **message length** (varint 1 to 10 bytes): is the length of the entire message in bytes
+- **compact share**: a type of share that can accomodate multiple units. Currently, compact shares are used for transactions, ISRs, and evidence to efficiently pack this information into as few shares as possible.
+- **sparse share**: a type of share that can accomodate zero or one unit. Currently, sparse shares are used for messages.
+
+## Context
+
+The current compact share format is:
+
+- First share of reserved namespace: <br>`nid (8 bytes) | reserved (1 byte) | data length (varint 1 to 10 bytes) | data`
+- Contiguous share in reserved namespace:<br>`nid (8 bytes) | reserved (1 byte) | data`
+
+The current sparse share format is:
+
+- First share of message:<br>`nid (8 bytes) | message length (varint 1 to 10 bytes) | data`
+- Contiguous share in message:<br>`nid (8 bytes) | data`
+
+The current share format poses multiple challenges:
+
+1. Clients must have two share parsing implementations (one for compact shares and one for sparse shares).
+1. It is difficult to make changes to the share format in a backwards compatible way because clients can't determine which version of the share format an individual share conforms to.
+1. It is not possible for a client that samples a random share to determine if the share is the first share of a message or a contiguous share in the message.
+
+## Proposal
+
+Introduce a universal share encoding that applies to both compact and sparse shares:
+
+- First share of namespace for compact shares or message for sprase shares:<br>`nid (8 bytes) | info (1 byte) | data length (varint 1 to 10 bytes) | data`
+- Contiguous share in namespace for compact shares or message for sparse shares:<br>`nid (8 bytes) | info (1 byte) | data`
+
+Shares in the reserved namespace have the added constraint: the first byte of `data` is a reserved byte so the format is:<br>`nid (8 bytes) | info (1 byte) | data length (varint 1 to 10 bytes) | reserved (1 byte) | data`
```

Comment:
[clarification] `data length (varint 1 to 10 bytes)` is the sum of the length of all units in a particular compact share namespace (i.e. all transactions in the transaction namespace). This is distinct the from transaction length varint that is prefixed to an individual transaction.

If so https://github.com/celestiaorg/celestia-app/pull/691 doesn't currently handle this.

Reply:
`data length (varint 1 to 10 bytes)` is confusing b/c for compact shares it is the sum of lengths for all units in a particular namespace but for messages, it is the length of a message which isn't necessarily the sum of lengths of all messages in the namespace.

Reply:
I think the trick is that conceptually we think of all the data in a reserved namespace as a single message.

Reply:
Created https://github.com/celestiaorg/celestia-app/issues/725 as a follow-up but great point @musalbas , will improve wording in this ADR based on that